src/dynamic_programming/matrix_chain_multiply.rs

```diff
@@ -1,76 +1,74 @@
+//! number of multiplications required for matrix chain multiplication.
+//!
+//! # Time Complexity
+//!
+//! The algorithm runs in O(n^3) time complexity and O(n^2) space complexity, where n is the
+//! number of matrices.
 
-pub fn matrix_chain_multiply(matrices: Vec<u32>) -> u32 {
-    let n = matrices.len();
-    if n <= 2 {
-        // No multiplications required.
+/// Calculates the minimum number of scalar multiplications required to multiply a chain
+/// of matrices with given dimensions.
+///
+/// # Arguments
+///
+/// * `dimensions`: A vector where each element represents the dimensions of consecutive matrices
+///   in the chain. For example, [1, 2, 3, 4] represents matrices of dimensions (1x2), (2x3), and (3x4).
+///
+/// # Returns
+///
+/// The minimum number of scalar multiplications needed to compute the product of the matrices
+/// in the optimal order.
+pub fn matrix_chain_multiply(dimensions: Vec<usize>) -> usize {
+    if dimensions.len() <= 2 {
         return 0;
     }
-    let mut table = vec![vec![0; n]; n];
 
-    for length in 2..n {
-        for i in 0..n - length {
-            let j = i + length;
-            table[i][j] = u32::MAX;
-            for k in i + 1..j {
-                let multiplications =
-                    table[i][k] + table[k][j] + matrices[i] * matrices[k] * matrices[j];
-                if multiplications < table[i][j] {
-                    table[i][j] = multiplications;
-                }
-            }
-        }
-    }
+    let mut min_operations = vec![vec![0; dimensions.len()]; dimensions.len()];
 
-    table[0][n - 1]
+    (2..dimensions.len()).for_each(|chain_len| {
+        (0..dimensions.len() - chain_len).for_each(|start| {
+            let end = start + chain_len;
+            min_operations[start][end] = (start + 1..end)
+                .map(|split| {
+                    min_operations[start][split]
+                        + min_operations[split][end]
+                        + dimensions[start] * dimensions[split] * dimensions[end]
+                })
+                .min()
+                .unwrap_or(usize::MAX);
+        });
+    });
+
+    min_operations[0][dimensions.len() - 1]
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    #[test]
-    fn basic() {
-        assert_eq!(matrix_chain_multiply(vec![1, 2, 3, 4]), 18);
-        assert_eq!(matrix_chain_multiply(vec![4, 3, 2, 1]), 18);
-        assert_eq!(matrix_chain_multiply(vec![40, 20, 30, 10, 30]), 26000);
-        assert_eq!(matrix_chain_multiply(vec![1, 2, 3, 4, 3]), 30);
-        assert_eq!(matrix_chain_multiply(vec![1, 2, 3, 4, 3]), 30);
-        assert_eq!(matrix_chain_multiply(vec![4, 10, 3, 12, 20, 7]), 1344);
+    macro_rules! test_cases {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (input, expected) = $test_case;
+                    assert_eq!(matrix_chain_multiply(input), expected);
```

Comment:
If you would like to _generate_ more test cases, you can do something like that:

```suggestion
                    assert_eq!(matrix_chain_multiply(input.clone()), expected);
                    assert_eq!(matrix_chain_multiply(input.into_iter().rev().collect()), expected);
```

Reply:
I will try it.

Reply:
All the tests passed.